File: leptos-kanban/src/types/state/all_boards.rs
use leptos::prelude::*;

use crate::types::{serialize::{AllBoards, BoardId, ColumnType, TaskId}, state::{BoardState, TaskState}};

#[derive(Clone, Copy, Debug)]
pub struct AllBoardsState {
    pub boards: RwSignal<Vec<BoardState>>,
    pub current_board_id: RwSignal<Option<BoardId>>,
}

impl AllBoardsState {
    pub fn from_all_boards(all_boards: AllBoards) -> Self {
        all_boards.into()
    }

    pub fn current_board(&self) -> Option<BoardState> {
        self.boards.with(|boards| {
            boards
                .iter()
                .find(|board| Some(board.id) == self.current_board_id.get())
                .copied()
        })
    }

    pub fn set_current_board(&self, board_id: BoardId) {
        self.current_board_id.set(Some(board_id));
    }

    pub fn add_board(&self, board: BoardState) {
        self.boards.update(|boards| boards.push(board));
        self.set_current_board(board.id);
    }

    pub fn add_task_to_current_board(&self, column_type: &ColumnType, task: TaskState) {
        if let Some(board) = self.current_board() {
            board.add_task_to_column(column_type, task);
        }
    }

    pub fn delete_task_from_current_board(&self, task_id: &TaskId) {
        if let Some(board) = self.current_board() {
            board.delete_task(task_id);
        }
    }

    pub fn move_task_in_current_board(
        &self,
        task_id: &TaskId,
        to_column_type: &ColumnType,
        before_task_id: Option<TaskId>,
    ) -> bool {
        self.current_board()
            .map(|board| board.move_task(task_id, to_column_type, before_task_id))
            .unwrap_or(false)
    }
}

impl From<AllBoards> for AllBoardsState {
    fn from(all_boards: AllBoards) -> Self {
        let serialized_current_board_id = all_boards.current_board_id();
        let board_states: Vec<BoardState> =
            all_boards.boards().iter().map(BoardState::from).collect();

        let current_board_id = serialized_current_board_id
            .filter(|current_id| board_states.iter().any(|board| board.id == *current_id))
            .or_else(|| board_states.first().map(|board| board.id));

        Self {
            boards: RwSignal::new(board_states),
            current_board_id: RwSignal::new(current_board_id),
        }
    }
}

File: react-kanban/src/types/state/all_boards.ts
import type {
  AllBoards,
  Board,
  BoardId,
  ColumnId,
  ColumnType,
  Task,
  TaskId,
} from "../serialize";
import {
  addTaskToBoardColumn,
  canMoveBoardTask,
  deleteBoardTask,
  moveBoardTask,
  updateBoardTask,
} from "./board";

export function getCurrentBoard(allBoards: AllBoards): Board | null {
  return (
    allBoards.boards.find((board) => board.id === allBoards.currentBoardId) ??
    null
  );
}

export function setCurrentBoard(
  allBoards: AllBoards,
  boardId: BoardId,
): AllBoards {
  return {
    ...allBoards,
    currentBoardId: boardId,
  };
}

export function addBoardToAllBoards(
  allBoards: AllBoards,
  board: Board,
): AllBoards {
  return {
    boards: [...allBoards.boards, board],
    currentBoardId: board.id,
  };
}

export function updateAllBoardsBoardTitle(
  allBoards: AllBoards,
  boardId: BoardId,
  title: string,
): AllBoards {
  return {
    ...allBoards,
    boards: allBoards.boards.map((board) =>
      board.id === boardId ? { ...board, title } : board,
    ),
  };
}

export function updateCurrentBoardColumnWipLimit(
  allBoards: AllBoards,
  columnId: ColumnId,
  wipLimit: number | null,
): AllBoards {
  return updateCurrentBoard(allBoards, (board) => ({
    ...board,
    columns: board.columns.map((column) =>
      column.id === columnId ? { ...column, wipLimit } : column,
    ),
  }));
}

export function addTaskToCurrentBoard(
  allBoards: AllBoards,
  columnType: ColumnType,
  task: Task,
): AllBoards {
  return updateCurrentBoard(allBoards, (board) =>
    addTaskToBoardColumn(board, columnType, task),
  );
}

export function updateTaskOnCurrentBoard(
  allBoards: AllBoards,
  taskId: TaskId,
  task: Task,
): AllBoards {
  return updateCurrentBoard(allBoards, (board) =>
    updateBoardTask(board, taskId, task),
  );
}

export function deleteTaskFromCurrentBoard(
  allBoards: AllBoards,
  taskId: TaskId,
): AllBoards {
  return updateCurrentBoard(allBoards, (board) =>
    deleteBoardTask(board, taskId),
  );
}

export function canMoveTaskInCurrentBoard(
  allBoards: AllBoards,
  taskId: TaskId,
  toColumnType: ColumnType,
  beforeTaskId: TaskId | null,
): boolean {
  const board = getCurrentBoard(allBoards);

  return board
    ? canMoveBoardTask(board, taskId, toColumnType, beforeTaskId)
    : false;
}

export function moveTaskInCurrentBoard(
  allBoards: AllBoards,
  taskId: TaskId,
  toColumnType: ColumnType,
  beforeTaskId: TaskId | null,
): AllBoards {
  return updateCurrentBoard(allBoards, (board) =>
    moveBoardTask(board, taskId, toColumnType, beforeTaskId),
  );
}

function updateCurrentBoard(
  allBoards: AllBoards,
  updateBoard: (board: Board) => Board,
): AllBoards {
  let changed = false;
  const boards = allBoards.boards.map((board) => {
    if (board.id !== allBoards.currentBoardId) {
      return board;
    }

    const nextBoard = updateBoard(board);
    changed ||= nextBoard !== board;
    return nextBoard;
  });

  return changed ? { ...allBoards, boards } : allBoards;
}
